samples/nrf9160/nrf_cloud_coap_client/src/main.c

```diff
@@ -0,0 +1,803 @@
+	cell_info.gci_cells = gci_cells;
+
+	LOG_INF("Requesting neighbor cell measurement");
+	err = lte_lc_neighbor_cell_measurement(&ncellmeas_params);
+	if (err) {
+		LOG_ERR("Failed to start neighbor cell measurement, error: %d", err);
+	} else {
+		request_cells = false;
+		LOG_INF("Waiting for measurement results...");
+	}
+}
+
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+static int do_pgps(struct gps_pgps_request *pgps_req)
+{
+	int err;
+	struct nrf_cloud_rest_pgps_request pgps_request;
+	struct nrf_cloud_pgps_result pgps_res;
+	static char host[64];
+	static char path[128];
+
+	LOG_INF("******** Getting P-GPS data");
+	memset(&pgps_request, 0, sizeof(pgps_request));
+	memset(&pgps_res, 0, sizeof(pgps_res));
+	pgps_request.pgps_req = pgps_req;
+	LOG_INF("period_min:%u, pred_cnt:%u, gps_day:%u, gps_time:%u",
+		pgps_req->prediction_period_min, pgps_req->prediction_count,
+		pgps_req->gps_day, pgps_req->gps_time_of_day);
+	memset(host, 0, sizeof(host));
+	memset(path, 0, sizeof(path));
+	pgps_res.host = host;
+	pgps_res.host_sz = sizeof(host);
+	pgps_res.path = path;
+	pgps_res.path_sz = sizeof(path);
+	err = nrf_cloud_coap_pgps_url_get(&pgps_request, &pgps_res);
+	if (err) {
+		LOG_ERR("Failed to request P-GPS: %d", err);
+		return err;
+	}
+	LOG_INF("P-GPS host:%s, host_sz:%u, path:%s, path_sz:%u",
+		pgps_res.host, pgps_res.host_sz, pgps_res.path, pgps_res.path_sz);
+	err = nrf_cloud_pgps_update(&pgps_res);
+	if (err) {
+		nrf_cloud_pgps_request_reset();
+		LOG_ERR("P-GPS data processing failed, error: %d", err);
+		return err;
+	}
+
+	LOG_DBG("P-GPS data processed");
+	err = nrf_cloud_pgps_notify_prediction();
+	if (err) {
+		LOG_ERR("GNSS: Failed to request current prediction, error: %d", err);
+	} else {
+		LOG_DBG("P-GPS prediction requested");
+	}
+	return err;
+}
+#endif
+
+#if defined(CONFIG_NRF_CLOUD_AGPS)
+static uint8_t agps_buf[4096];
+
+static int do_agps(void)
+{
+	int err;
+	static bool got_agps;
+	struct nrf_cloud_rest_agps_request agps_request;
+	struct nrf_modem_gnss_agps_data_frame agps_req;
+	struct nrf_cloud_rest_agps_result agps_res;
+
+	if (got_agps) {
+		LOG_INF("Not required yet.");
+		return 0;
+	}
+	memset(&agps_request, 0, sizeof(agps_request));
+	memset(&agps_req, 0, sizeof(agps_req));
+	agps_request.type = NRF_CLOUD_REST_AGPS_REQ_CUSTOM;
+	agps_request.net_info = &cell_info;
+	agps_request.agps_req = &agps_req;
+	agps_req.data_flags = 0x3f;
+	agps_req.sv_mask_alm = 0xffffffff;
+	agps_req.sv_mask_ephe = 0xffffffff;
+	agps_res.buf = agps_buf;
+	agps_res.buf_sz = sizeof(agps_buf);
+	err = nrf_cloud_coap_agps_data_get(&agps_request, &agps_res);
+	if (err) {
+		LOG_ERR("Failed to request A-GPS: %d", err);
+	} else {
+		LOG_INF("Processing %u bytes of A-GPS data", agps_res.agps_sz);
+		err = nrf_cloud_agps_process(agps_res.buf, agps_res.agps_sz);
+		if (err) {
+			LOG_ERR("A-GPS data processing failed, error: %d", err);
+		}  else {
+			LOG_INF("A-GPS data processed");
+			got_agps = true;
```

Comment:
this is never set back to false, so A-GPS is only done once in the sample's lifetime?

Reply:
Yes, I have not decided whether I should implement the full set of GNSS interactions with the modem, including handling the assistance needed event, or just wait for the location library update.